vllm/engine/arg_utils.py

```diff
@@ -1162,6 +1165,58 @@ def create_engine_config(self) -> VllmConfig:
         )
 
 
+@dataclass
+class EngineArgsV1(_EngineArgs):
+    """Arguments for vLLM engine v1."""
+
+    # V1's default values that differ from the default values in EngineArgs.
+    # This allows to switch between V1 and V0's default behaviour transparently.
+    enable_prefix_caching: bool = True
+
+    @staticmethod
+    def add_cli_args(parser: FlexibleArgumentParser) -> FlexibleArgumentParser:
+        parser = _EngineArgs.add_cli_args(parser)
+        return parser
+
+    def create_engine_config(self,
+                             usage_context: Optional[UsageContext] = None
+                             ) -> VllmConfig:
+        assert (usage_context
+                is not None), "usage_context must be provided for EngineArgsV1"
+
+        if self.max_num_batched_tokens is None:
+            # When no user override, set the default values based on the
+            # usage context.
+            if usage_context == UsageContext.LLM_CLASS:
+                logger.warning("Setting max_num_batched_tokens to 8192 "
+                               "for LLM_CLASS usage context.")
+                self.max_num_seqs = 1024
+                self.max_num_batched_tokens = 8192
+            elif usage_context == UsageContext.OPENAI_API_SERVER:
+                logger.warning("Setting max_num_batched_tokens to 2048 "
+                               "for OPENAI_API_SERVER usage context.")
+                self.max_num_seqs = 1024
+                self.max_num_batched_tokens = 2048
+
+        engine_config = super().create_engine_config(usage_context)
+
+        # TODO (ywang96): Enable APC by default when VLM supports it.
+        if engine_config.model_config.is_multimodal_model:
+            logger.warning(
+                "Prefix caching is currently not supported for multimodal "
+                "models and has been disabled.")
+            engine_config.cache_config.enable_prefix_caching = False
+        return engine_config
+
+
+EngineArgs = _EngineArgs  # type: ignore
+
+if envs.VLLM_USE_V1:
+    # Overwrite EngineArgs to use EngineArgsV1
+    # This has to be done before `AsyncEngineArgs` is imported.
+    EngineArgs = EngineArgsV1  # type: ignore
```

Comment:
dynamically changing the class looks quite strange. for example, if someone wants to create both a v0 engine args and v1 engine args for testing, it will not be possible under this PR.

I think you can have v1 config override inside the `create_engine_config()` function, and read `envs.VLLM_USE_V1` there.

Reply:
> for example, if someone wants to create both a v0 engine args and v1 engine args for testing, it will not be possible under this PR.

Yeah, I think one will have to reimport the file with `VLLM_USE_V1=0/1` to do this. But I am not sure how niche this usecase would be. 

> I think you can have v1 config override inside the create_engine_config() function, and read envs.VLLM_USE_V1 there.

Yeah, I think the only issue with this is we want to have different default value for v1 and v0. But maybe there's some way to make this possible too. Let me try.